Email Editor - Move custom editor to the post editor page [MAILPOET-6090]

costasovo opened the pull request:
## Description

This PR moves the email editor to `wp-admin/post.php?post={postId}&action=edit` page and removes the custom page.

With this change, it might be easier to switch to a post editor in the future, and in case we use native WP's listing for posts, the edit link will go to the correct editor.

## Code review notes
I kept all the code in MailPoet plugin integration in `MailPoetmailEditor\Integrations\MailPoetditorPageRenderer`. I think it makes sense for now because the assets we enqueue are from the plugin. In the future, we should move some of that code to the PHP package to make things easier for an integrator. 

## QA notes

_N/A_

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-6090]

## After-merge notes

_N/A_

## Tasks

- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-6090]: https://mailpoet.atlassian.net/browse/MAILPOET-6090?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Preview

[Preview in WordPress Playground](https://account.mailpoet.com/playground/new/branch:email-editor/move-to-post-page)

_The latest successful build from `email-editor/move-to-post-page` will be used. If none is available, the link won't work._

costasovo committed:
Render custom editor for mailpoet_email post types

[MAILPOET-6090]

costasovo committed:
Improve detection that we are on the email editor page to cover also post editor

This is needed so that we can register some early hooks.
[MAILPOET-6090]

costasovo committed:
Update links to the email editor to go to post.php

[MAILPOET-6090]

costasovo committed:
Highlight the MailPoet > Emails menu item when editing email post

[MAILPOET-6090]

costasovo committed:
Remove the custom editor page

[MAILPOET-6090]

costasovo committed:
Improve the comment about checking the editor page

[MAILPOET-6090]

triple0t commented on mailpoet/lib/EmailEditor/Integrations/MailPoet/EmailEditor.php:
```diff
@@ -65,4 +77,14 @@ public function extendEmailPostApi() {
+
+  public function replaceEditor($replace, $post) {
+    $currentScreen = get_current_screen();
+    if (use_block_editor_for_post($post) && $post->post_type === self::MAILPOET_EMAIL_POST_TYPE && $currentScreen) { // phpcs:ignore Squiz.NamingConventions.ValidVariableName.MemberNotCamelCaps
```
[Nitpick]
Since we are already checking for the MailPoet custom post type. Do we still need to use the `use_block_editor_for_post`?  
On checking the [docs](https://developer.wordpress.org/reference/functions/use_block_editor_for_post/), the function has a filter that can be used to disable the block editor for certain (or all post type), is that a behavior we expect and want?

costasovo replied:
> On checking the [docs](https://developer.wordpress.org/reference/functions/use_block_editor_for_post/), the function has a filter that can be used to disable the block editor for certain (or all post type), is that a behavior we expect and want?

Nice catch! We want to always use block editor so the `use_block_editor_for_post` makes no sense. Removed in https://github.com/mailpoet/mailpoet/pull/5930/commits/475ef6d9df01f18c046d6a5cbab8bc4d7560e696

triple0t commented on mailpoet/assets/js/src/newsletters/campaign-stats/newsletter-stats-info.tsx:
```diff
@@ -38,9 +38,8 @@ const getEditorLink = (newsletter: NewsletterType) => {
     newsletter.wp_post_id
   ) {
-    editorHref = `admin.php?page=mailpoet-email-editor&postId=${newsletter.wp_post_id}`;
+    editorHref = `post.php?post=${newsletter.wp_post_id}&action=edit`;
```
We missed [this](https://github.com/mailpoet/mailpoet/blob/7bcfe93269b3532056cf5f6878aa70bff614af0d/mailpoet/assets/js/src/newsletters/send.tsx#L876-L879) check.  We might also use this opportunity to consider the initial option you raised of using a custom function for the Gutenberg email editor URL

costasovo replied:
I would bet I was searching codebase for `mailpoet-email-editor`. Somehow I missed one. Thanks a lot for double checking. I added a function for the Gutenberg Editor URL in https://github.com/mailpoet/mailpoet/pull/5930/commits/912fe668233c26082f3bca4eb027f5dbefa54271

triple0t commented on mailpoet/lib/Config/Menu.php:
```diff
@@ -257,27 +261,6 @@ private function registerMailPoetSubMenuEntries() {
-
-    // Add body class for email editor page
-    // We need to mark the page as a block editor page so that some of the block editor styles are applied properly
-    $this->wp->addAction('load-' . $emailEditorPage, function() {
-      $this->wp->addFilter('admin_body_class', function ($classes) {
-        return ltrim($classes . ' block-editor-page');
-      });
-    });
```
We removed the `block-editor-page` class here, but I discovered that the new email editor page does not contain this class.  
![Screenshot 2024-11-15 at 11 28 54 AM](https://github.com/user-attachments/assets/614bb3e8-6929-4680-87c5-14df01c4815f)

Noting it down in case it's important.

costasovo replied:
@triple0t  The class `block-editor-page` should be present on the body element. I can see it there. 
![Screenshot 2024-11-15 at 15 37 58](https://github.com/user-attachments/assets/4c7ce6b6-c9ba-419c-ae15-262d7c2a9bf5). It causes visual issues when missing.

I can see on your screenshot that you are focused in `editor-canvas` iframe. Please see just above the console command in your screenshot. This would explain why the query selector didn't find it. Could you please double check?

triple0t replied:
I just checked again, and I can confirm the presence of the CSS class. Thank you for pointing out the console.log context

triple0t reviewed (commented):
This looks good overall. Thank you for working on this. I added some comments, kindly check them below.

Final question from me: Do you think we should consider adding a redirect or something to the new URL if a link like this `/wp-admin/admin.php?page=mailpoet-email-editor&postId=POST_ID` already exists in the wild?

costasovo committed:
Simplify check for replacing editor

We always want to use the block email editor for mailpoet_email CPT.
We don't need to check if it is allowed for the CPT.
[MAILPOET-6090]

costasovo committed:
Introduce and use a function for block editor page

This way, we don't need to search for the URL and edit it in multiple places.
[MAILPOET-6090]

costasovo commented:
@triple0t Thank you for the thorough review! I'm happy we caught the issues. They should be addressed in new commits. Could you please take another look?

> Final question from me: Do you think we should consider adding a redirect or something to the new URL if a link like this /wp-admin/admin.php?page=mailpoet-email-editor&postId=POST_ID already exists in the wild?

I wouldn't do that. The email editor is behind a feature flag. Technically, it is possible for users to enable it, but even the page with feature flags is kind of secret and not linked from anywhere.  I don't expect many people to use it, and even if so, it is experimental functionality, so changes like that are expected. My thinking here is that we added the feature flags so that we are allowed to change things this way. Does that make sense to you?

triple0t commented:
> I wouldn't do that. The email editor is behind a feature flag. Technically, it is possible for users to enable it, but even the page with feature flags is kind of secret and not linked from anywhere. I don't expect many people to use it, and even if so, it is experimental functionality, so changes like that are expected. My thinking here is that we added the feature flags so that we are allowed to change things this way. Does that make sense to you?

Noted. That's fine with me. Well done, @costasovo

triple0t reviewed (approved):
LGTM
